pkg/virt-operator/resource/apply/prometheus_test.go

```diff
@@ -0,0 +1,203 @@
+
+		r := &Reconciler{
+			kv:           kv,
+			stores:       stores,
+			clientset:    clientset,
+			expectations: expectations,
+		}
+
+		requiredSM := sm.DeepCopy()
+		updatedEndpoints := []promv1.Endpoint{
+			{
+				Port: "metrics-update",
+			},
+		}
+		requiredSM.Spec.Endpoints = updatedEndpoints
+
+		patched := false
+		promClient.Fake.PrependReactor("patch", "servicemonitors", func(action testing.Action) (handled bool, ret runtime.Object, err error) {
+			a := action.(testing.PatchActionImpl)
+			patch, err := jsonpatch.DecodePatch(a.Patch)
+			Expect(err).ToNot(HaveOccurred())
+
+			patched = true
+
+			obj, err := json.Marshal(sm)
+			Expect(err).To(BeNil())
+
+			obj, err = patch.Apply(obj)
+			Expect(err).To(BeNil())
+
+			sm := &promv1.ServiceMonitor{}
+			Expect(json.Unmarshal(obj, sm)).To(Succeed())
+			Expect(sm.Spec.Endpoints).To(Equal(updatedEndpoints))
+
+			return true, sm, nil
+		})
+
+		Expect(r.createOrUpdateServiceMonitor(requiredSM)).To(BeNil())
+		Expect(patched).To(BeTrue())
+	})
+
+	It("should not patch PrometheusRules on sync when they are equal", func() {
+
+		pr := components.NewPrometheusRuleCR("namespace", config.WorkloadUpdatesEnabled())
+
+		version, imageRegistry, id := getTargetVersionRegistryID(kv)
+		injectOperatorMetadata(kv, &pr.ObjectMeta, version, imageRegistry, id, true)
+
+		stores.PrometheusRuleCache.Add(pr)
+
+		r := &Reconciler{
+			kv:           kv,
+			stores:       stores,
+			clientset:    clientset,
+			expectations: expectations,
+		}
+
+		Expect(r.createOrUpdatePrometheusRule(pr)).To(BeNil())
+	})
+
+	It("should patch PrometheusRules on sync when they are equal", func() {
+
+		pr := components.NewPrometheusRuleCR("namespace", config.WorkloadUpdatesEnabled())
+
+		version, imageRegistry, id := getTargetVersionRegistryID(kv)
+		injectOperatorMetadata(kv, &pr.ObjectMeta, version, imageRegistry, id, true)
+
+		stores.PrometheusRuleCache.Add(pr)
+
+		r := &Reconciler{
+			kv:           kv,
+			stores:       stores,
+			clientset:    clientset,
+			expectations: expectations,
+		}
+
+		requiredPR := pr.DeepCopy()
+		updatedGroups := []promv1.RuleGroup{
+			{
+				Name: "Updated",
+			},
+		}
+		requiredPR.Spec.Groups = updatedGroups
+
+		patched := false
+		promClient.Fake.PrependReactor("patch", "prometheusrules", func(action testing.Action) (handled bool, ret runtime.Object, err error) {
+			a := action.(testing.PatchActionImpl)
+			patch, err := jsonpatch.DecodePatch(a.Patch)
+			Expect(err).ToNot(HaveOccurred())
+
+			patched = true
+
+			obj, err := json.Marshal(pr)
+			Expect(err).To(BeNil())
+
+			obj, err = patch.Apply(obj)
+			Expect(err).To(BeNil())
+
+			pr := &promv1.PrometheusRule{}
+			Expect(json.Unmarshal(obj, pr)).To(Succeed())
+			Expect(pr.Spec.Groups).To(Equal(updatedGroups))
+			Expect(len(pr.Spec.Groups)).To(Equal(1))
+
+			return true, pr, nil
+		})
+
+		Expect(r.createOrUpdatePrometheusRule(requiredPR)).To(BeNil())
+		Expect(patched).To(BeTrue())
+	})
+})
```

Comment:
@ashleyschuett It seems there is a lot of common code between the different "It" functions, I think that you can try using  table.DescribeTable and consolidate the different cases into just table entries.  I didn't look too deep into what it takes to do it - so feel free to ignore if you don't think it will help.